src/wp-includes/class-wp-query.php

```diff
@@ -3528,8 +3529,17 @@ public function get_posts() {
 			/** @var WP_Post[] */
 			$this->posts = array_map( 'get_post', $this->posts );
 
-			if ( $q['cache_results'] ) {
-				update_post_caches( $this->posts, $post_type, $q['update_post_term_cache'], $q['update_post_meta_cache'] );
+			if ( $q['cache_results'] && $id_posts_are_cacheable ) {
+				update_post_cache( $this->posts );
+			}
+
+			if ( $q['update_post_meta_cache'] ) {
+				$post_ids = wp_list_pluck( $this->posts, 'ID' );
```

Comment:
See my other comments. It's a bit cheaper to use a simple `foreach`. It's 2-3 more lines of code, but that's okay. :)